### PR TITLE
Bump localVersion upon navigation and test it

### PR DESCRIPTION
**Pull request by alshakero**
Fixes: https://github.com/Palindrom/Palindrom/issues/217

### CODE REVIEW

**Comment by alshakero** [comment]
The failing test is flaky and is irrelevant to this PR.

**Comment by miyconst** [comment]
Even though the proposed solution "bumps" the local version upon navigation, it does not resolve the problem. What currently happens, is:

1.1. Palindrom bumps local version and notifies about it the server side via an empty patch.
1.2. Palindrom sends an HTTP request to perform navigation.
2.1. Server receives the empty patch to bump version.
2.2. Server receives the HTTP request to perform navigation.
3.1. Server responds to the empty patch.
3.2. Server responds to the HTTP request.
4.1. Palindrom fires `onPatchReceived` callback.
4.2. Palindrom fires `palindrom-after-navigation` event. If the proposed [fix](https://github.com/Palindrom/Palindrom/issues/217#issuecomment-485448200) is applied.

The problem here, is that due to the network nature, all `N.1` and `N.2` items will be executed in parallel, with undefined order between them.

Thus, the `onPatchReceived` callback caused by a navigation, may be called **before** and **after** the actual navigation.

The issue also does not solve any consistency problems, when navigation happens before already queued patches applied.

The proper solution should send an HTTP `POST`/`PUT` request with request body, which bumps the local version.

I don't think, that the issue can be fixed on client side only, it's client-server feature, thus I also don't think we shall merge current solution, as it does not solve the problem, but rather brings more confusion.

Instead, I would rather just add another callback, something like `onAfterNavigation`.

**Comment by alshakero** [comment]
> 1.1. Palindrom bumps local version and notifies about it the server side via an empty patch.

I don't think this is true. The call I added don't send anything to the server. It just bumps a JS var. Is your review based on testing or just reading the code? Because I realize the code is a little misleading.

**Comment by miyconst** [comment]
>I don't think this is true. The call I added doesn't send anything to the server. It just bumps a JS var. Is your review based on testing or just reading the code? Because I realize the code is a little misleading.

It was based on reading the code. 

If this is the case, means, if the server does not even get to know about the client side bump, then how does the server confirm, that the bumped client side version is applied?

**Comment by alshakero** [comment]
> if the server does not even get to know about the client side bump, then how does the server confirm, that the bumped client side version is applied?

The server does know. It implies the bump from the HTTP PATCH request and acts accordingly. And it confirmation is also implicit when the server responds.

File: test/specs/dom/ot.js [hunk 1]
```diff
@@ -134,5 +134,49 @@ if (typeof window !== 'undefined') {
...
+            contentType: 'application/json-patch+json'
+          },
+          responseText: JSON.stringify(patch1)
+        });
+        
+        palindrom.morphUrl(url2);
+
+        setTimeout(() => {
+          // make sure the version is bumped now
+          assert.equal(palindrom.queue.localVersion, 1);
```
Comment (warpech) [inline test/specs/dom/ot.js]: Can you add an assertion that checks if there was zero communication on WebSocket? The patch should only take exactly one HTTP request and response.

**Comment by miyconst** [comment]
@alshakero and me had a Slack call and came to a conclusion, that:

- Bumping local version and sending an HTTP request shall be a single transaction.
- Server side shall be able to identify on which client side version to perform the navigation.
- Thus, the bumped client side version has to be supplied to the server with the HTTP request.
- Since currently we perform navigation with HTTP `GET`, the client version shall be supplied as a custom header.

Overall, this issue is more complex than it seemed to be, thus @alshakero most likely won't be able to do anything for it.

**Comment by warpech** [comment]
There is a problem, but slightly different than @miyconst says. 

The line: `this.palindrom.queue.send([]);` bumps the local version, but the new version number is never sent to the server.

1. The version number is not sent in the HTTP GET request for the new URL. This used to be incorrect to send a HTTP GET with a body (see: https://stackoverflow.com/questions/978061/http-get-with-request-body). Now it is not formally incorrect, however `XmlHttpRequest` will not send the body anyway, [per spec](https://xhr.spec.whatwg.org/#the-send()-method):

> `If the request method is GET or HEAD, set body to null.`

2. The version number should not be sent in WebSocket either, because that's a different communication channel.

------

I think that the only sensible solution is what @alshakero proposes:

> The server does know. It implies the bump from the HTTP PATCH request and acts accordingly. And it confirmation is also implicit when the server responds.

The server should assume that any HTTP GET with `application/json-patch+json` bumps the client version.

Right now, the server ignores version numbers anyway (because of the missing server-side OT https://github.com/Starcounter/Home/issues/453).

-----

My conclution is that the implementation in this PR is good, and it is compatible with the current Starcounter server.

**Comment by miyconst** [comment]
>The server **should** assume that any HTTP GET with `application/json-patch+json` bumps the client version. But in my manual testing it looks that it is not assuming that, and it is holding the OT queue infinitely after a navigation to a new URL.

There is a problem with this solution: server **does not know** what was the previous client side version prior to the navigation.

If there are 10 patches on the client side queue, which are being sent over the web socket, the HTTP GET navigation request may arrive to the server at any moment of time, and there is no way for the server to know, whether it has to apply the navigation *now* or wait for the pending patches to arrive over the web socket first.

Unfortunately, current solution does not solve the problem, thus I see no reason to merge it.

**Comment by warpech** [comment]
You are very right. I wrote my comment before reading https://github.com/Palindrom/Palindrom/pull/232#issuecomment-487492412. 

Then I think our options are:

1. keep sending the HTTP GET request with `Accepts: application/json-patch+json` to the app URL, just include the new version number as a header
2. add a special meta field in the session JSON for the application URL (`_url$`) that will be used for redirections to a new URL. This removes the need to ever send a HTTP GET for anything else than a first request, will use WebSocket with HTTP PATCH fallback.

**Comment by miyconst** [comment]
I think we shall align this with @erikvk as well. Anyway, both options have pros and cons, and even though I would love to go for the option 2, I don't see how we can overcome it's cons.

---------

>1. Keep sending the HTTP GET request with `Accepts: application/json-patch+json` to the app URL, just include the new version number as a header.

- ✔️ Less server side changes.
- ✔️ Easier to implement.
- ✔️ Cookies support. ⭐
- ❌ Custom HTTP header.
- ❌ Multiple communication channels.
- ❌ Possible code duplication between regular patch and navigation request implementation.

---------

>2. Add a special meta field in the session JSON for the application URL (_url$) that will be used for redirections to a new URL. This removes the need to ever send a HTTP GET for anything else than a first request, will use WebSocket with HTTP PATCH fallback.

- ✔️ One communication channel for both patch types.
- ✔️ Possibly less code duplication.
- ❌ More server side changes.
- ❌ Custom view-model property, which assumes *there is a URL*. Not applicable for WPF clients, for example.
- ❌ No easy way to set cookies on navigation. ⭐

---------

The cookie feature is a deal breaker for me.

**Comment by erikvk** [comment]
> The cookie feature is a deal breaker for me.
@miyconst please explain

**Comment by miyconst** [comment]
@erikvk, HTTP cookie is a special HTTP header. 

The request / response headers are only supported for HTTP requests. 

For example, when Sign In app needs to set session key cookie, it has to either be a response to an HTTP request, or utilize some custom made mechanism, to deliver cookie value over the web socket and set it on the client.

**Comment by erikvk** [comment]
@miyconst OK, and we need to allow cookies to be set on navigation requests? Since that could possibly navigate to a different app?

**Comment by miyconst** [comment]
>@miyconst OK, and we need to allow cookies to be set on navigation requests? Since that could possibly navigate to a different app?

Yes.

**Comment by erikvk** [comment]
OK cool. I'll just give my 2 C on this.

Option 2 is common to how I have implemented navigation in [RESTar WebSocket resources](https://develop.mopedo.com/RESTar/Consuming%20a%20RESTar%20API/Consuming%20terminal%20resources/). In RESTar, there is a meta object, being a representation of the session itself, holding metadata properties like uptime, navigation, versions, headers etc. Option 2 is essentially the same idea, but the url property being flattened to the parent object.

The problem with headers is handled by there being a global header set for each websocket connection, that is populated with headers from the websocket upgrade GET request. The header collection is then patchable as well, being a member of the meta object. I haven't tried it with session cookies however.

**Comment by erikvk** [comment]
With that being said, I think option 2 + a mechanism for working with headers is clearly the way to go. But there might be palindrom-specific things I'm not aware of.

**Comment by warpech** [comment]
Let's get back to this soon. Meanwhile, merging of this PR is on hold.

**Comment by warpech** [comment]
I just thought about a variant of option 1, in which we don't create a custom HTTP header for the version number, but a custom HTTP header for the whole patch body.

In this solution, we can bypass the limitation of not being able to send a request payload in body, we will send the full request payload in the header, like:

```
GET /people/person/123 HTTP/1.1
Accept: application/json-patch+json
X-Palindrom-Patch: [{"op": "test", "path": "/_ver#s", "value": 0}, {"op": "replace", "path": "/_ver#c$", "value": 1}]
```

**Comment by erikvk** [comment]
> I just thought about a variant of option 1, in which we don't create a custom HTTP header for the version number, but a custom HTTP header for the whole patch body.
> 
> In this solution, we can bypass the limitation of not being able to send a request payload in body, we will send the full request payload in the header

I've implemented similar features in the past, to get around this very limitation. It may generate a bunch of issues, where header value encoding gets messed up or misinterpreted. For example, the web server used in Starcounter 2.3 interprets all line breaks in header values as separations of headers. Back when I tried this last, I ended up requiring clients to base64 encode header values containing arbitrary characters. So maybe we would need something like that here too.

**Comment by miyconst** [comment]
The proper implementation would still be to use `PUT` or `POST` methods and post the patch in the request body.

This issue is gonna be discussed between @dan31, @erikvk, and me during this week.

**Comment by warpech** [comment]
> The proper implementation would still be to use `PUT` or `POST` methods and post the patch in the request body.

You are right, or actually, the most proper implementation would be to use the `PATCH` method to the handler URI. Then we can send regular patches to the handler with no workarounds. The only requirement is to change `Handle.GET` to actually handle `GET` and PATCH` requests.

**Comment by alshakero** [comment]
Closing as obselete. Please create a new PR if you'd like to continue.

**Pull request closed without merging**